### PR TITLE
Fix several Chomp issues

### LINKED ISSUE #1452: chomp/OptimizerAdapter: Start state is empty
### Description
Attempted to use `chomp/OptimizerAdapter` after `OMPL`. Followed tutorials and faced the problem of `Start state is empty`. The example print-out can be seen at the bottom,

### Your environment
* ROS Distro: Melodic
* OS Version: Ubuntu 18.04
* Binary build, with `moveit_planners` built from source
* `ros-melodic-moveit` binaries `1.0.1-0bionic.20190320.192509`
* `moveit_planners` from `melodic-devel`

### Steps to reproduce

Start the tutorial through the **Getting Started** portion of [MoveIt! Tutorials](http://docs.ros.org/kinetic/api/moveit_tutorials/html/doc/getting_started/getting_started.html).

After setting up the workspace, start following the steps of [Planning Adapter Tutorials](http://docs.ros.org/kinetic/api/moveit_tutorials/html/doc/planning_adapters/planning_adapters_tutorial.html), under **Running OMPL as a pre-processor for CHOMP**.

I noticed that there might be some errors in this section, in order to use this setup with `chomp/OptimizerAdapter`, we would need the `ompl` planner to be called instead of `chomp`, so I believe an error we should be launching
```
roslaunch panda_moveit_config dmeo.launch
```
instead of 
```
roslaunch panda_moveit_config demo_chomp.launch
```

The next error I got was 
```
[ERROR] [1556590982.275248450]: Exception while loading planning adapter plugin 'chomp/OptimizerAdapter': According to the loaded plugin descriptions the class chomp/OptimizerAdapter with base class type planning_request_adapter::PlanningRequestAdapter does not exist. Declared types are  default_planner_request_adapters/AddIterativeSplineParameterization default_planner_request_adapters/AddTimeOptimalParameterization default_planner_request_adapters/AddTimeParameterization default_planner_request_adapters/Empty default_planner_request_adapters/FixStartStateBounds default_planner_request_adapters/FixStartStateCollision default_planner_request_adapters/FixStartStatePathConstraints default_planner_request_adapters/FixWorkspaceBounds industrial_trajectory_filters/AddSmoothingFilter industrial_trajectory_filters/NPointFilter industrial_trajectory_filters/UniformSampleFilter
```

After digging around, I found that the `chomp/OptimizerAdapter` was not built in the binaries, so I did the following
```
cd /tmp
git clone -b melodic-devel https://github.com/ros-planning/moveit
cp -r /tmp/moveit/moveit_planners ~/ws_moveit/src/.
```
Rebuilding them again now,
```
cd ~/ws_moveit
catkin clean -y
catkin build
```
I then got a dependency issue, which was solved with
```
sudo apt-get install ros-melodic-moveit-resources
catkin build
```
The workspace built without any issues, however when I launched the demo via
```
roslaunch panda_moveit_config demo.launch
```
adding the **Motion Planning** panel in `rviz`, under **Planning** tab -> **Planning Group**: panda_arm ->**Goal State**: random_valid, **Plan and Execute**. I got the following logs

### Console output and the files I edited from tutorials
https://gist.github.com/aaronchongth/116e85a65d26e3daef7c47fbed98e209

### PR DESCRIPTION
**Pull request by rhaschke**
This PR comprises several independent commits to improve the Chomp code, namely
- some code cleanup, reducing redundancy and explicit loops
- fix start state handling: 
  - An empty start state should be accepted. In this case, the current state should be used. Fixes #1452
  - Checking start state constraints was plain wrong, resulting in #1470.
- Avoid extra conversion from/to MotionPlanResponse _message_

### CODE REVIEW

**Comment by BryceStevenWilley** [comment]
Looks good on a brief scan, I'll have some time to take a look at this PR this weekend.

File: moveit_planners/chomp/chomp_interface/test/chomp_moveit_test.cpp [hunk 1]
```diff
@@ -88,14 +88,6 @@ TEST_F(CHOMPMoveitTest, cartesianGoal)
   EXPECT_EQ(error_code.val, moveit::planning_interface::MoveItErrorCode::INVALID_GOAL_CONSTRAINTS);
 }
 
-TEST_F(CHOMPMoveitTest, noStartState)
```
Comment (bmagyar) [inline moveit_planners/chomp/chomp_interface/test/chomp_moveit_test.cpp]: How are invalid start states handled now?

File: moveit_planners/chomp/chomp_motion_planner/include/chomp_motion_planner/chomp_planner.h [hunk 2]
```diff
@@ -38,21 +38,21 @@
...
 #include <moveit/planning_scene/planning_scene.h>
 
 namespace chomp
 {
 class ChompPlanner
 {
 public:
-  ChompPlanner();
-  virtual ~ChompPlanner(){};
+  ChompPlanner() = default;
```
Comment (bmagyar) [inline moveit_planners/chomp/chomp_motion_planner/include/chomp_motion_planner/chomp_planner.h]: Is this even worth keeping is we don't do anything in it?

File: moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp [hunk 3]
```diff
@@ -43,64 +43,58 @@
...
     ROS_ERROR_STREAM("Only joint-space goals are supported");
-    res.error_code.val = moveit_msgs::MoveItErrorCodes::INVALID_GOAL_CONSTRAINTS;
+    res.error_code_.val = moveit_msgs::MoveItErrorCodes::INVALID_GOAL_CONSTRAINTS;
     return false;
   }
 
-  int goal_index = trajectory.getNumPoints() - 1;
-  trajectory.getTrajectoryPoint(goal_index) = trajectory.getTrajectoryPoint(0);
-  sensor_msgs::JointState js;
+  size_t goal_index = trajectory.getNumPoints() - 1;
```
Comment (bmagyar) [inline moveit_planners/chomp/chomp_motion_planner/src/chomp_planner.cpp]: didn't see any other uses for it
```suggestion
  const size_t goal_index = trajectory.getNumPoints() - 1;
```

File: moveit_planners/chomp/chomp_motion_planner/src/chomp_trajectory.cpp [hunk 4]
```diff
@@ -297,27 +229,30 @@ bool ChompTrajectory::fillInFromTrajectory(moveit_msgs::MotionPlanDetailedRespon
   {
     // perform a decimation sampling in this block if the number of trajectory points in the MotionPlanDetailedResponse
     // res object is more than the number of points in the CHOMP trajectory
-    double decimation_sampling_factor = ((double)num_response_points) / ((double)num_chomp_trajectory_points);
+    double decimation_sampling_factor = ((double)num_input_points) / ((double)num_chomp_trajectory_points);
```
Comment (bmagyar) [inline moveit_planners/chomp/chomp_motion_planner/src/chomp_trajectory.cpp]:
```suggestion
    const double decimation_sampling_factor = ((double)num_input_points) / ((double)num_chomp_trajectory_points);
```

File: moveit_planners/chomp/chomp_motion_planner/src/chomp_trajectory.cpp [hunk 5]
```diff
@@ -231,64 +169,58 @@ void ChompTrajectory::fillInMinJerk()
...
+  if (num_input_points < 2)
     return false;
 
-  // get the number of joints for each robot state
-  int num_joints_trajectory = trajectory_msg.joint_trajectory.points[0].positions.size();
-
   // variables for populating the CHOMP trajectory with correct number of trajectory points
-  int repeated_factor = num_chomp_trajectory_points / num_response_points;
-  int repeated_balance_factor = num_chomp_trajectory_points % num_response_points;
+  unsigned int repeated_factor = num_chomp_trajectory_points / num_input_points;
```
Comment (bmagyar) [inline moveit_planners/chomp/chomp_motion_planner/src/chomp_trajectory.cpp]:
```suggestion
  const unsigned int repeated_factor = num_chomp_trajectory_points / num_input_points;
```

File: moveit_planners/chomp/chomp_motion_planner/src/chomp_trajectory.cpp [hunk 6]
```diff
@@ -231,64 +169,58 @@ void ChompTrajectory::fillInMinJerk()
...
+  unsigned int repeated_balance_factor = num_chomp_trajectory_points % num_input_points;
```
Comment (bmagyar) [inline moveit_planners/chomp/chomp_motion_planner/src/chomp_trajectory.cpp]:
```suggestion
  const unsigned int repeated_balance_factor = num_chomp_trajectory_points % num_input_points;
```

File: moveit_planners/chomp/chomp_motion_planner/src/chomp_trajectory.cpp [hunk 7]
```diff
@@ -231,64 +169,58 @@ void ChompTrajectory::fillInMinJerk()
...
+bool ChompTrajectory::fillInFromTrajectory(const robot_trajectory::RobotTrajectory& trajectory)
 {
-  // create a RobotTrajectory msg to obtain the trajectory from the MotionPlanDetailedResponse object
-  moveit_msgs::RobotTrajectory trajectory_msg = res.trajectory[0];
-
   // get the default number of points in the CHOMP trajectory
-  int num_chomp_trajectory_points = (*this).getNumPoints();
-  // get the number of points in the response trajectory obtained from OMPL
-  int num_response_points = trajectory_msg.joint_trajectory.points.size();
+  size_t num_chomp_trajectory_points = (*this).getNumPoints();
```
Comment (bmagyar) [inline moveit_planners/chomp/chomp_motion_planner/src/chomp_trajectory.cpp]:
```suggestion
  const size_t num_chomp_trajectory_points = (*this).getNumPoints();
```

File: moveit_planners/chomp/chomp_motion_planner/src/chomp_trajectory.cpp [hunk 8]
```diff
@@ -231,64 +169,58 @@ void ChompTrajectory::fillInMinJerk()
...
+  // get the number of points in the input trajectory
+  size_t num_input_points = trajectory.getWayPointCount();
```
Comment (bmagyar) [inline moveit_planners/chomp/chomp_motion_planner/src/chomp_trajectory.cpp]:
```suggestion
  const size_t num_input_points = trajectory.getWayPointCount();
```

File: moveit_planners/chomp/chomp_motion_planner/include/chomp_motion_planner/chomp_planner.h (hunk 2 above)
Comment (rhaschke) [reply moveit_planners/chomp/chomp_motion_planner/include/chomp_motion_planner/chomp_planner.h]: Otherwise, the default constructor wouldn't be created.

**Comment by rhaschke** [comment]
I addressed your comments, @bmagyar. Thanks a lot. Indeed, it's always a good idea to declare const variables.

File: moveit_planners/chomp/chomp_motion_planner/include/chomp_motion_planner/chomp_planner.h (hunk 2 above)
Comment (bmagyar) [reply moveit_planners/chomp/chomp_motion_planner/include/chomp_motion_planner/chomp_planner.h]: I think it should. When you don't declare it at all, you get the default constructor. When you have one declared, the compiler won't generate the default one for you but the `default` keyword can be used to force generating it. Unfortunately I don't have a setup available ATM where I could double check it but I'm pretty certain.

The story about the destructor is slightly different as if you declare it virtual you are allowing derived classes to do a proper cleanup.

Comment (rhaschke) [reply moveit_planners/chomp/chomp_motion_planner/include/chomp_motion_planner/chomp_planner.h]: I believe the default constructor is only created by default if there are no other constructors defined. But, I might be wrong. Nevertheless, it's not wrong to explicitly declare that we want to have the default constructor :wink:

Comment (bmagyar) [reply moveit_planners/chomp/chomp_motion_planner/include/chomp_motion_planner/chomp_planner.h]: Yeah and as far as I see this is the only constructor being defined. This conversation is dragging for too long, let's go to merge :D

File: moveit_planners/chomp/chomp_interface/test/chomp_moveit_test.cpp (hunk 1 above)
Comment (bmagyar) [reply moveit_planners/chomp/chomp_interface/test/chomp_moveit_test.cpp]: Thanks

**Comment by bmagyar** [comment]
@BryceStevenWilley , next time you are around and happy with the PR, feel free to merge :+1:

**Pull request merged**